### ISSUE TITLE
Training large datasets with natural gradients

### ISSUE THREAD

**Issue by theogf** [GPflow/GPflow#820]
Hello,

I tried to use the natural gradient method using the example in the notebook in 
doc/source/notebooks/natural_gradients.ipynb
When creating some random data via sklearn.datasets : make_classification, the algorithm works nicely for small datasets (<1000 samples) but going higher going higher leads to the error : 

`InvalidArgumentError (see above for traceback): Cholesky decomposition was not successful. The input might not be valid.
	 [[Node: NatGradOptimizer/Cholesky_1 = Cholesky[T=DT_DOUBLE, _device="/job:localhost/replica:0/task:0/device:CPU:0"](NatGradOptimizer/mul_6)]]`

For information I kept the same values as in the notebook, namely \gamma = 0.1 for the natural gradient updates and lr = 0.01 for the Adam updates.
Could you provide a solution in order to make the algorithm stable? 
Thanks.

**Comment by hughsalimbeni** [comment]
Sometimes a \gamma value of 0.1 is too large in the non-conjugate case. One option is to increases \gamma in the first few iterations. Here is a (somewhat contrived) example where a fixed step of 0.1 is too big

```python
N, M = 2000, 100
X = np.linspace(-10, 10, N).reshape(-1, 1)
Z = np.linspace(-10, 10, M).reshape(-1, 1)
Y = np.zeros((N, 1))
Y[int(N/2):] = 1.
Y[int(3*N/4):] = 2.

model = SVGP(X, Y, RBF(1), MultiClass(3), Z=Z, num_latent=3)
model.q_mu.set_trainable(False)
model.q_sqrt.set_trainable(False)

op_ng_fixed = NatGradOptimizer(0.1).make_optimize_tensor(model, var_list=[[model.q_mu, model.q_sqrt]])

sess = model.enquire_session()

try:
    sess.run(op_ng_fixed)
except tf.errors.InvalidArgumentError:
    print('Cholesky failure!')
```
This fails.

Instead of just using a smaller value, we can define a simple linearly increasing schedule for \gamma:

```python
gamma_start = 1e-5
gamma_max = 1e-1
gamma_step = 1e-3

gamma = tf.Variable(gamma_start, dtype=tf.float64) 

# linear increase up to the max value:
gamma_incremented = tf.where(tf.less(gamma, gamma_max), gamma + gamma_step, gamma_max)

op_ng = NatGradOptimizer(gamma).make_optimize_tensor(model, var_list=[[model.q_mu, model.q_sqrt]])
op_adam = AdamOptimizer(0.001).make_optimize_tensor(model)
op_increment_gamma = tf.assign(gamma, gamma_incremented)

sess.run(tf.variables_initializer([gamma]))  # NB gpflow is handling all the other variable inits

for it in range(1000):
    try:
        sess.run(op_ng)
        sess.run(op_increment_gamma)
    except tf.errors.InvalidArgumentError:
        print('gamma = {} on iteration {} is too big!'.format(it, sess.run(gamma)))
        break  # we'll just stop if it fails
        
    sess.run(op_adam)

    if it % 100 == 0:
        print('{} gamma={:.5f} ELBO={:.4f}'.format(it, *sess.run([gamma, model.likelihood_tensor])))
```

The above code will still fail if the increase is too aggressive (e.g. starting from 0.01). To make things more safe, we can do something like the following

```python
gamma_start = 1e-2   # deliberately chosen to be too large for this example
gamma_max = 1e-1   # same max value as before 
gamma_step = 1e-2  # this is much more aggressive increase

gamma = tf.Variable(gamma_start, dtype=tf.float64)
gamma_incremented = tf.where(tf.less(gamma, gamma_max), gamma + gamma_step, gamma_max)

op_ng = NatGradOptimizer(gamma).make_optimize_tensor(model, var_list=[[model.q_mu, model.q_sqrt]])
op_adam = AdamOptimizer(0.001).make_optimize_tensor(model)
op_increment_gamma = tf.assign(gamma, gamma_incremented)

gamma_fallback = 1e-1   # we'll reduce by this factor if there's a cholesky failure 
op_fallback_gamma = tf.assign(gamma, gamma * gamma_fallback)  

sess.run(tf.variables_initializer([gamma]))

for it in range(1000):
    try:
        sess.run(op_ng)
        sess.run(op_increment_gamma)
    except tf.errors.InvalidArgumentError:
        g = sess.run(gamma)
        print('gamma = {} on iteration {} is too big! Falling back to {}'.format(it, g, g * gamma_fallback))
        sess.run(op_fallback_gamma)
        
    sess.run(op_adam)

    if it % 100 == 0:
        print('{} gamma={:.4f} ELBO={:.4f}'.format(it, *sess.run([gamma, model.likelihood_tensor])))
```

**Comment by theogf** [comment]
Thanks this definitely helped! However I still have a small question, is there a reason you use incremental steps : 
`gamma + gamma_step)` 
as opposed to multiplicative steps (i.e. `gamma*gamma_mult` , where (`gamma_mult>1`) ?

**Comment by hughsalimbeni** [comment]
[Here](https://arxiv.org/pdf/1803.09151.pdf) we did multiplicative increments, but now I tend to do linear (for no particular reason, but it seems to work fine also).

**Comment by st--** [comment]
@theogf closing this issue as it's not actually a bug in our code. If you'd like other people to benefit from what you've learnt, I would strongly suggest asking a question on stackoverflow.com with tag "gpflow" (https://stackoverflow.com/questions/tagged/gpflow), and answering it yourself! (You'll get some stackoverflow reputation out of it, too...) Thanks :)
